### ISSUE TITLE
'Error: WebSocket is not open' when closing two connected ws clients simultaneously

### ISSUE THREAD

**Issue by GrzegorzKazana** [marblejs/marble#312]
Hi 👋 
first of all, thanks for providing this awesome package - I am having great time using it. Unfortunately I ran into an issue, that I believe is not on my side (however I am happy to be proven wrong, or pointed to a valid solution).

**Describe the bug**
When closing connections of two clients that are connected and notified of each other state via a common channel (e.g. a Subject), an `Error: WebSocket is not open: readyState 2 (CLOSING)` is thrown. Happens only when clients are closed immediately after another (e.g. via a script).

Full stack trace:
```
Error: WebSocket is not open: readyState 2 (CLOSING)
    at WebSocket.send (/*path to the project*//node_modules/@marblejs/websockets/node_modules/ws/lib/websocket.js:322:19)
    at WebSocket.sendResponse (/*path to the project*//node_modules/@marblejs/websockets/dist/response/websocket.response.handler.js:7:12)
    at SafeSubscriber.input$.pipe.subscribe.type [as _next] (/*path to the project*//node_modules/@marblejs/websockets/dist/server/websocket.server.listener.js:64:42)
    at SafeSubscriber.__tryOrUnsub (/*path to the project*//node_modules/rxjs/src/internal/Subscriber.ts:265:10)
    at SafeSubscriber.next (/*path to the project*//node_modules/rxjs/src/internal/Subscriber.ts:207:14)
    at Subscriber._next (/*path to the project*//node_modules/rxjs/src/internal/Subscriber.ts:139:22)
    at Subscriber.next (/*path to the project*//node_modules/rxjs/src/internal/Subscriber.ts:99:12)
    at TakeUntilSubscriber.Subscriber._next (/*path to the project*//node_modules/rxjs/src/internal/Subscriber.ts:139:22)
    at TakeUntilSubscriber.Subscriber.next (/*path to the project*//node_modules/rxjs/src/internal/Subscriber.ts:99:12)
    at CatchSubscriber.Subscriber._next (/*path to the project*//node_modules/rxjs/src/internal/Subscriber.ts:139:22)
```

**To Reproduce**
1. Link two clients using a common channel.
2. Using the `finalize` operator notify each user about another user being disconnected
3. Open the connections
4. Close the connections directly one after another
(5.) First client that happens to be disconnected causes a message to be pushed to the second client, that also disconnected in the meantime.
(6.) Message is being pushed, even though the websocket is in `CLOSING` state

Minimal epic causing the error:
```ts
class UserPoolPubSub {
    private readonly pubSub = new Subject<Message>();
    public join(id: string): Observable<Message> {
        this.publish(id, 'Hello there!');
        return this.pubSub.pipe(filter(({ publisher }) => publisher !== id));
    }
    public publish(publisher: string, message: string): void {
        this.pubSub.next({ publisher, message });
    }
}

const pool = new UserPoolPubSub();

const join$: WsEffect = (event$, { client: { id } }) => {
    const leave$ = event$.pipe(matchEvent('LEAVE'));

    return event$.pipe(
        matchEvent('JOIN'),
        exhaustMap(() =>
            pool.join(id).pipe(
                map(({ message }) => ({ type: 'RECEIVED', payload: message })),
                takeUntil(leave$),
            ),
        ),
        finalize(() => pool.publish(id, 'I was disconnected, sorry')),
    );
};
```

Minimal client interaction:
```ts
function run() {
// after opening the ws connections and exchanging messages
  clientA.close();
  clientB.close();
}
```

I published a runnable reproduction here: https://github.com/GrzegorzKazana/marblejs-ws-issue-repro.
It can be run by
```
npm i
npm run dev
# in separate terminal
npm run client
```

**Expected behavior**
Server should not try to push the event to user that is already closing.

**Desktop (please complete the following information):**
 ```
OS: MacOS 10.14.6
@marblejs/core: 3.4.8
@marblejs/websockets: 3.4.8
Node version: 12.18.0, also happens on 14.15.1
```

**Additional context**
I run into this issue when doing integration tests for my app, that uses `@marblejs/websocket` for its backend. When the test runner was halting on a failed test, it caused closing all connections (2 in my case) at the same time - this in turned was causing crash of my backend service.

I hope I did not miss any obvious error on my side, but I will be happy to be corrected or pointed to a mistake.

Thanks in advance!

**Comment by JozefFlakus** [comment]
Hi, thanks for pointing out this issue.

I think the issue here [`#handleResponse`](https://github.com/marblejs/marble/blob/master/packages/websockets/src/response/websocket.response.handler.ts#L14) implementation. Basically the response handler has missing checks for client ready state `WebSocket.OPEN`. I thought it was handled directly in `ws` lib that the `@marblejs/websockets` module utilizes. IMO that's not a problem to addd it.

There is one thing that I'm only thinking about. How the server should react in case of different ready state.
1. `OPEN` is obvious...
2. `CLOSED` and `CLOSING` - maybe it would be a good idea to log a warning that there was an event trying to be sent.
3. `CONNECTING` server should wait till te ready state won't be `OPEN` - if `CLOSED` or `CLOSING should end waiting and log a message (see point 2).

WDYT?

**Comment by GrzegorzKazana** [comment]
Right, adding the checks would definitely solve the problem - I guess that creators of `ws` decided not to add them because of the ambiguity of how the server should react - and left that to the consumers of the library.

On how we should react, I mostly agree,
- `OPEN` is indeed obvious
- `CLOSING` should be a warning, nothing we can do about it
- `CONNECTING` it could either log the warning, or buffer the events. However I am not sure if buffering matches any use-case. If the client is just connecting, how would we know we need to push any data to it - the client did not have a chance to subscribe to any source. If I have not overlooked a valid use case, I do not think it is worth the additional complexity.
- `CLOSED` in theory we should be safe from this one (since [this](https://github.com/marblejs/marble/blob/f7e87f663b82ade350ab0d2083f017b8f9f367a9/packages/websockets/src/server/websocket.server.listener.ts#L116)), so I am not sure if we should bother with this case.

So in the end, I think that we would be fine with just logging a warning on every other case than `OPEN` (given that pushing to socket in `CONNECTING` state is very much unlikely and pushing to `CLOSED` socket _in theory_ should not happen).

**Comment by JozefFlakus** [comment]
I'm just trying to figure out where the `CONNECTING` state could happen in case of the mentioned response handler. Generally speaking it is impossible to reach that state because the handler reacts to the already connected clients.

>However I am not sure if buffering matches any use-case. If the client is just connecting, how would we know we need to push any data to it - the client did not have a chance to subscribe to any source [...]

Correct, I'm also not sure if it is worth to bother with this state, taking into consideration the fact that websocket is a full-duplex. Client cannot sent a message to the server when the connection is not established (handshake process is not completed). In this case it is up to the client how it will be buffered. 🤔

**Comment by GrzegorzKazana** [comment]
Btw, just out of curiosity, how would you go about changing the behavior of `handleResponse`? Broadly speaking, I see 3 options that would work, but would like to hear your opinion 👀 

1. Return `Observable<false>` when readyState is invalid, log it outside based on the return value
2. Return IOEither, fold it outside and log on `left`
3. Pass logger into `handleResponse` - `WebSocket -> EventTransformer -> Logger -> Event -> Observable<boolean>` (given we already make a side effect by doing `.send`, why not make another one by logging in the same place 😅 )
4. ???

**Comment by JozefFlakus** [comment]
@GrzegorzKazana Agree, I see that this part has to be rewritten ;)

**Comment by JozefFlakus** [comment]
In this case the tricky thing is that the `handleResponse` can be used either directly inside an effect (it is monkey patched to the `client.sendResponse` attribute) and server listener upon subscription to outgoing events. If the logging (side-effect) will be moved down (outside the response handler) then the direct **client.sendResponse** call won't inform developer that he/she is trying to send an event to an already closed/closing connection.

> given we already make a side effect by doing `.send`, why not make another one by logging in the same place
Taking into consideration the mentioned thing (☝️) IMO that's the most efficient way of doing it.

**Comment by GrzegorzKazana** [comment]
That's a really accurate remark - I completely overlooked the fact that `sendResponse` is exposed to the user (via the `client`). Guess we can not change the public api too much then - so the option 3. does in fact seem to be the optimal way of going about it 🚀

**Comment by JozefFlakus** [comment]
Version [`3.4.9`](https://github.com/marblejs/marble/releases/tag/v3.4.9) has been released 🚀
